File: gamesPlatform/Client/GenericPages/CanvasContext.cs
using System.Runtime.InteropServices.JavaScript;
using System.Runtime.Versioning;

namespace cmArcade.Client.GenericPages;

[SupportedOSPlatform("browser")]
internal partial class CanvasContext
{
    [JSImport("setup", "CanvasContext")]
    internal static partial void SetupContext();
    [JSImport("setupImage", "CanvasContext")]
    internal static partial void SetupImage();
    [JSImport("startBatch", "CanvasContext")]
    internal static partial void StartBatch(); 
    [JSImport("endBatch", "CanvasContext")]
    internal static partial void EndBatch();


    [JSImport("setStrokeStyle", "CanvasContext")]
    internal static partial void SetStrokeStyle(string style);
    [JSImport("stroke", "CanvasContext")]
    internal static partial void Stroke();
    [JSImport("setFillStyle", "CanvasContext")]
    internal static partial void SetFillStyle(string style);
    [JSImport("fill", "CanvasContext")]
    internal static partial void Fill();


    [JSImport("closePath", "CanvasContext")]
    internal static partial void ClosePath();
    [JSImport("beginPath", "CanvasContext")]
    internal static partial void BeginPath();
    [JSImport("moveTo", "CanvasContext")]
    internal static partial void MoveTo(float x, float y);
    [JSImport("lineTo", "CanvasContext")]
    internal static partial void LineTo(float x, float y);
    [JSImport("fillRect", "CanvasContext")]
    internal static partial void FillRect(float x, float y, float width, float height);
    [JSImport("clearRect", "CanvasContext")]
    internal static partial void ClearRect(float x, float y, float width, float height);
    [JSImport("forceClear", "CanvasContext")]
    internal static partial void ForceClear(float x, float y, float width, float height);


    [JSImport("fillText", "CanvasContext")]
    internal static partial void FillText(string text, float x, float y);
    [JSImport("setFont", "CanvasContext")]
    internal static partial void SetFont(string font);


    [JSImport("drawImage", "CanvasContext")]
    internal static partial void DrawImage(string imgID, 
        float? sx, float? sy, float? sWidth, float? sHeight, 
        float dx, float dy, float? dWidth, float? dHeight);

    internal static void DrawImage(string imgID, float dx, float dy, float? dWidth, float dHeight) =>
        DrawImage(imgID, null, null, null, null, dx, dy, dWidth, dHeight);

    internal static void DrawImage(string imgID, float dx, float dy) =>
        DrawImage(imgID, null, null, null, null, dx, dy, null, null);
}

File: gamesPlatform/Client/wwwroot/CanvasContext.js
﻿let currentContext = undefined;
let currentCanvas = undefined;

let drawingContext = undefined;

export function setup() {
    currentCanvas = document.getElementById("game-display-canvas");
    currentContext = currentCanvas.getContext("2d");
    drawingContext = currentContext;
}

export function fillText(msg, x, y) {
    drawingContext.fillText(msg, x, y);
}

export function setFont(font) {
    drawingContext.font = font;
}

export function setStrokeStyle(style) {
    drawingContext.strokeStyle = style;
}

export function stroke() {
    drawingContext.stroke();
}

export function setFillStyle(style) {
    drawingContext.fillStyle = style;
}

export function fill() {
    drawingContext.fill();
}

export function closePath() {
    drawingContext.closePath();
}

export function beginPath() {
    drawingContext.beginPath();
}

export function moveTo(x, y) {
    drawingContext.moveTo(x, y);
}

export function lineTo(x, y) {
    drawingContext.lineTo(x, y);
}

export function fillRect(x, y, width, height) {
    drawingContext.fillRect(x, y, width, height);
}

export function clearRect(x, y, width, height) {
    drawingContext.clearRect(x, y, width, height);
}

export function clearCanvas() {
    if (currentContext)
        currentContext.clearRect(0, 0, currentCanvas.width, currentCanvas.height);
}

export function drawImage(imageID, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight) {
    const imageElement = document.getElementById("im-" + imageID);

    if (dWidth !== null && dHeight !== null) {
        if (sx !== null && sy !== null)
            drawingContext.drawImage(imageElement, sx, sy, sWidth, sHeight, dx, dy, dWidth, dHeight);
        else
            drawingContext.drawImage(imageElement, dx, dy, dWidth, dHeight);
    } else
        drawingContext.drawImage(imageElement, dx, dy);
}

export function startBatch() {
    currentCanvas.offscreenCanvas = new OffscreenCanvas(currentCanvas.width, currentCanvas.height);
    drawingContext = currentCanvas.offscreenCanvas.getContext("2d");
}

export function endBatch() {
    currentContext.drawImage(currentCanvas.offscreenCanvas, 0, 0);
    currentCanvas.offscreenCanvas = null;
    drawingContext = currentContext;
}